### ISSUE TITLE
[Question] Sentry Integration for error logging

### ISSUE THREAD

**Issue by khoerintus** [temporalio/sdk-php#103]
Is there any best practice to integrate temporal with sentry for error logging?
I know that I can use sentry php sdk, try catch in my code, and send the error.
But is there any global error handler that catch every error (from Workflow or Activity), so I can just send the error from there?
Or I can integrate roadrunner with sentry?

**Comment by khoerintus** [comment]
Thanks @SerafimArts for the answer. I will try it later.

**Comment by khoerintus** [comment]
Hi @SerafimArts , I've tried to write my own transport decorator and use it. But the error function is never called.

I found that Exception thrown in Workflow class, is caught in Temporal\Internal\Workflow\Process\start(),
and Exception thrown in Activity class, is caught in Temporal\Internal\Transport\Router\handle().
If I put the code to send the error to sentry in those functions, it sent successfully.

**Comment by SerafimArts** [comment]
Yes, that's right. The example I showed only intercepts fatal SDK errors, and all "business logic" errors are sent directly to the Temporal.

Hmmm... @mfateev perhaps it makes sense to add APM (Sentry, New Relic, etc) support to the Temporal Server?

**Comment by changwuf31** [comment]
Meanwhile, what is the current solution if we want to use sentry for temporal and php sdk ?

**Comment by wolfy-j** [comment]
None. We are still thinking how would that look like. Most likely we can provide you hook entry point for activities, but not for the workflows as I don't think Sentry SDK can work in async mode.

**Comment by kozlice** [comment]
I ran into the same issue, except I have multiple handlers in monolog (file log + Sentry). Doing a try-catch with logging & re-throwing inside an activity is not very convenient.

The proposal document mentions the idea of injecting a PSR-3-compatible logger into a worker: https://github.com/temporalio/proposals/blob/master/php/php-rr-sdk.md#logging Was it rejected or it's just not implemented yet?

**Comment by wolfy-j** [comment]
Not implemented due to the fact that you create activity classes manually, so you control injections to them.

The container refactor planned later this year, with some internal changes and optimizations, including the ability to intercept activity call (for exception catching or state resets).

**Comment by wolfy-j** [comment]
You can find an example of integration here: https://github.com/spiral/temporal-bridge/blob/master/src/InvokeActivityRouter.php

Essentially you can intercept the activity call and manipulate it with input or output parameters, or handle exceptions.

**Comment by changwuf31** [comment]
@wolfy-j :+1:  Superb, however it will be much better, if some of the sdk properties are made protected, so we can simply override the necessary functions.

**Comment by programmador** [comment]
The link above is broken as that functionality was removed from temporal-bridge.
Anyway, [here](https://github.com/spiral/temporal-bridge/commit/2de00229c26e35a7cb83d7e6027270464ecbb996#diff-c6622530d00ec62cd9f77427ade1d89488623a20917ffa1066323dab91e460a9)'s how it looked like.
Though fully copying the whole WorkerFactory instead of just extending is redundant as for me. The only reason was to access `$this->queues` which is private. Accessing parent's private field is tricky, I've embraced [this library](https://github.com/ivastly/php-reflection).

And yes, this is really-really hacky, "don't do that at home".

BTW You may ask why was that functionality removed from temporal-bridge. The reason is introducing "finalizers" functionality which is enough for example for cleaning resources to avoid memory leaks but not enough for error handling as finalizers are working similar to `final` block after `try-catch`. So You cannot access errors in finalizers.
